whittle/pretrain_super_network.py

```diff
@@ -0,0 +1,505 @@
+        print(f"Available values:\n{available_models}")
+        quit()
+
+    if initial_checkpoint_dir is not None:
+        initial_checkpoint_dir = extend_checkpoint_dir(initial_checkpoint_dir)
+
+    if tokenizer_dir is not None:
+        tokenizer_dir = extend_checkpoint_dir(tokenizer_dir)
+
+    if model_config is None:
+        # Support both model_name options: meta-llama/Meta-Llama-3-8B & Meta-Llama-3-8B
+        try:
+            model_config = Config.from_name(model_name)
+        except ValueError:
+            print(f"Model name {model_name} is not supported.\n")
+            available_models = "\n".join(sorted(name_to_config))
+            print(f"Available values:\n{available_models}")
+            quit()
+
+    assert training_strategy in training_strategies_cls, print(
+        f"Training strategy is {training_strategy}. Should be in {list(training_strategies_cls)}"
+    )
+
+    hparams = capture_hparams()
+    data = TinyLlama() if data is None else data
+
+    config = Config.from_name(model_name) if model_config is None else model_config
+    config.fix_head_size = True
+
+    precision = precision or get_default_supported_precision(training=True)
+    num_devices = parse_devices(devices)
+    out_dir = init_out_dir(out_dir)
+    # in case the dataset requires the Tokenizer
+    tokenizer = Tokenizer(tokenizer_dir) if tokenizer_dir is not None else None
+
+    logger = choose_logger(
+        logger_name,
+        out_dir,
+        name=f"pretrain-{config.name}",
+        resume=bool(resume),
+        log_interval=train.log_interval,
+    )
+
+    if num_devices * num_nodes > 1:
+        strategy = FSDPStrategy(
+            auto_wrap_policy={Block},
+            state_dict_type="full",
+            sharding_strategy="HYBRID_SHARD",
+        )
+    else:
+        strategy = "auto"
+
+    fabric = L.Fabric(
+        devices=num_devices,
+        num_nodes=num_nodes,
+        strategy=strategy,
+        precision=precision,
+        loggers=[logger],
+    )
+
+    if torch.cuda.is_available() and num_devices > 1:
+        check_nvlink_connectivity(fabric)
+
+    fabric.launch()
+
+    fabric.print(pprint.pformat(hparams))
+    if logger_name in ("tensorboard", "wandb"):
+        fabric.logger.log_hyperparams(hparams)
+
+    main(
+        fabric,
+        num_devices,
+        seed,
+        initial_checkpoint_dir,
+        resume,
+        config,
+        data,
+        out_dir,
+        tokenizer_dir,
+        tokenizer,
+        train,
+        eval,
+        optimizer,
+        training_strategy,
+    )
+
+
+def fit(
+    fabric: L.Fabric,
+    devices: int,
+    state: dict,
+    train_dataloader: DataLoader,
+    val_dataloader: DataLoader,
+    out_dir: Path,
+    tokenizer_dir: Optional[Path],
+    train: TrainArgs,
+    eval: EvalArgs,
+    training_strategy: BaseTrainingStrategy,
+) -> None:
+    model = state["model"]
+    optimizer = state["optimizer"]
+
+    if eval.initial_validation:
```

Comment:
Do we want to simply use litgpt's validation loop? Do we want to reset super network before every validation? Ideally would be nice to report : smallest, medium and largest architecture loss for a more robust validation

Reply:
You are right we should definitely reset the super-network. I am not sure about evaluating the smallest, medium and largest arch, though. I agree that it is convenient for diagnosis, but it also adds an additional overhead. Maybe we could have different callbacks to switch between different validation themes? Could be a separate PR.